name_tracker.go

```diff
@@ -6,14 +6,47 @@ package quamina
 // Quamina's MatchesForEvent API is strongly related to the number of fields it has to look at.  Therefore, a
 // Flattener should call NameTracker for each field it encounters and if it returns false, skip that field
 // and not add it to the flattened output. Here's a clarifying example; consider this JSON event:
-//  {"a": {"b": 1, "c": 2}}
+//
+//	{"a": {"b": 1, "c": 2}}
+//
 // Assuming NameTracker always returns positive, the flattened output should look like
-//  "a\nb" 1
-//  "a\nc" 2
+//
+//	"a\nb" 1
+//	"a\nc" 2
+//
 // Note that the values stored in NameTracker are not the full paths "a\nb" and "a\nc", but the path segments
 // "a", "b", and "c".  This reduces the work the Flattener has to do - whenever it processes an object, it need
 // only look up the member names in the NameTracker. A little thought reveals that this can produce false
 // positives and some potential wasted work, but is a good trade-off.
 type NameTracker interface {
 	IsNameUsed(label []byte) bool
+
+	// This should be actually part of matcher interface and not NameTracker.
+	GetHierarchalNameTracker() HierarchalNameTracker
+}
+
+type HierarchalNameTracker interface {
```

Comment:
I think it would be nice to have a long-ish comment here introducing what this type exists to do and why it has this design.

I'm also wondering about the `Name` in the type name. JSON brings the concept of "member name" and then Quamina has the concept of "Path" and then "segment" (which is actually just a member name). Is it possible to give this type a name based on these existing comments?  Or perhaps a name based on what useful thing it actually **does** e.g. "pathSkipper" or some such?

Reply:
Talked about it here - https://github.com/timbray/quamina/pull/124#issuecomment-1277377837

About a name, what it's actually doing in practice is "allowing to efficiently select an hierarchical data", so "pathSkipper" don't think it's a good name, what about `SegmentsTreeTracker` ? (I'm really bad a this =X)

Reply:
Added a comment an renamed - interface to `SegmentsTreeTracker` and `pathsIndex` to `segmentsTree`, if you have a better names I'll replace it :)